clang/include/clang/Basic/Diagnostic.td

```diff
@@ -155,6 +155,38 @@ class DefaultWarnNoWerror {
 }
 class DefaultRemark { Severity DefaultSeverity = SEV_Remark; }
 
+// C++ compatibility warnings.
+multiclass CXXCompat<
+    string message,
+    int std_ver,
+    bit ext_warn = true,
+    string std_ver_override = ""#std_ver> {
+    // 'X is a C++YZ extension'.
+    def compat_pre_cxx#std_ver#_#NAME :
+        Diagnostic<!strconcat(message, " a C++", std_ver_override,  " extension"),
+                   CLASS_EXTENSION,
+                   !if(ext_warn, SEV_Warning, SEV_Ignored)>,
+        InGroup<!cast<DiagGroup>("CXX"#std_ver)>;
+
+    // 'X is incompatible with C++98' (if std_ver == 11).
+    // 'X is incompatible with C++ standards before C++YZ' (otherwise).
+    def compat_cxx#std_ver#_#NAME :
+        Warning<!if(!eq(std_ver, 11),
+                    !strconcat(message, " incompatible with C++98"),
+                    !strconcat(message, " incompatible with C++ standards before C++", std_ver_override))>,
+        InGroup<!cast<DiagGroup>(!if(!eq(std_ver, 11),
+                                     "CXX98Compat",
+                                     "CXXPre"#std_ver#"Compat"))>,
+        DefaultIgnore;
+}
+
+multiclass CXX11Compat<string message, bit ext_warn = true> : CXXCompat<message, 11, ext_warn>;
```

Comment:
It would help to have leading comments that say:
```
// Generates compat_cxx11_<diag_name> and compat_pre_cxx11_<diag_name> diagnostic IDs.
```
so it's more clear what the name of the diagnostic will be.

Note, we usually have a convention that extension warnings are prefixed with `ext`, warnings are prefixed with `warn`; should we retain that convention?

Reply:
> Note, we usually have a convention that extension warnings are prefixed with `ext`, warnings are prefixed with `warn`; should we retain that convention?

Well, I currently at least always have to think for like half a minute when issuing a compatibility warning along the lines of ‘so if this is a C++20 feature, then this is a warning about C++17 compatibility, meaning it’s an extension in C++17, so it’s the C++17 warning that starts with `ext_`’. Imo that doesn’t really add too much value and just increases cognitive load (at least that’s how it is for me ;Þ). Also, with compatibility warnings, I don’t think we care too much to know at the call site whether it’s an ExtWarn/Extension or Warning.

Reply:
The reason it's beneficial is because it helps the reader understand "this code is an extension" vs "this code is standard functionality". We could also accomplish that with the helper function call rather than the diagnostic name itself, though.

Reply:
@AaronBallman I feel like having the helper would basically solve the problem I’m describing, so I can add `ext_`/`warn_` back to the warnings (but I’d prefer to do that after introducing the helper so I don’t have to go and rename the warnings everywhere only to change those callsites *again* in the follow-up pr ;Þ)

Reply:
The helper function would address my concerns such that we probably don't need to rename the diagnostics at all.